In [2]:
import pandas as pd

import numpy as np

import seaborn as sns

import matplotlib.pyplot as plt

import datetime as dt

In [3]:
data = pd.read_excel('data/26.3 Когортный анализ.xlsx')

In [4]:
data.head()

user_id       date  price
0   A0005 2020-03-25     50
1   A0005 2020-03-25     40
2   A0005 2020-03-26     70
3   A0005 2020-03-27     20
4   A0005 2020-03-29     50

In [5]:
def get_month(x):
    return dt.datetime(x.year, x.month, 1)

In [6]:
data['month_year'] = data.date.apply(get_month)

In [8]:
data['cohort_month'] = data.groupby('user_id').month_year.transform("min")

In [9]:
data.head()

user_id       date  price month_year cohort_month
0   A0005 2020-03-25     50 2020-03-01   2020-03-01
1   A0005 2020-03-25     40 2020-03-01   2020-03-01
2   A0005 2020-03-26     70 2020-03-01   2020-03-01
3   A0005 2020-03-27     20 2020-03-01   2020-03-01
4   A0005 2020-03-29     50 2020-03-01   2020-03-01

In [10]:
def get_date(df, column):
    year = df[column].dt.year
    month = df[column].dt.month
    day = df[column].dt.day
    return year, month, day 

In [11]:
data_year, data_month, _ = get_date(data, 'date')


In [12]:
cohort_year, cohort_month, _ = get_date(data, 'cohort_month')


In [13]:
year_diff = data_year - cohort_year
month_diff = data_month - cohort_month

In [14]:
data['cohort_index'] = year_diff * 12 + month_diff + 1

In [28]:
data_id = data.pivot_table(index="cohort_month",
                         columns='cohort_index',
                         values='user_id',
                         aggfunc=len)

In [29]:
first_column = data_id.iloc[:, 0]

relation_id = data_id.div(first_column, axis=0)

In [30]:
data_sum = data.pivot_table(index='cohort_month',
                            columns='cohort_index',
                            values='price',
                            aggfunc=sum)

C:\Users\lorex\AppData\Local\Temp\ipykernel_15076\553677987.py:1: FutureWarning: The provided callable <built-in function sum> is currently using DataFrameGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  data_sum = data.pivot_table(index='cohort_month',


In [46]:
first_column_sum = data_sum.iloc[:, 0]
relation_sum = data_sum.div(first_column_sum, axis=0)

In [41]:
display(data_id)
display(data_sum)


cohort_index       1        2       3       4       5       6       7   \
cohort_month                                                             
2020-01-01     9994.0   8175.0  7099.0  6637.0  6171.0  5127.0  4657.0   
2020-02-01     8463.0   8013.0  5565.0  5593.0  4952.0  4184.0  3293.0   
2020-03-01    13210.0   9779.0  8148.0  7837.0  7410.0  6000.0  5634.0   
2020-04-01    10123.0   8725.0  6272.0  6019.0  5769.0  4106.0  3699.0   
2020-05-01    14982.0  10806.0  8616.0  8336.0  7728.0  6609.0  5956.0   
2020-06-01    11304.0   9783.0  6793.0  6603.0  6294.0  5182.0  4710.0   
2020-07-01    17007.0  11428.0  8432.0  8273.0  7204.0  6673.0  6218.0   
2020-08-01    13458.0  10518.0  7241.0  7070.0  6935.0  5753.0  1038.0   
2020-09-01    17781.0  12152.0  8846.0  8656.0  7829.0  1493.0     NaN   
2020-10-01    15039.0  11141.0  7957.0  7422.0  1547.0     NaN     NaN   
2020-11-01    18887.0  12610.0  8804.0  1824.0     NaN     NaN     NaN   
2020-12-01    16582.0  12460.0  1638.0     NaN     NaN     NaN     NaN   
2021-01-01    20532.0   3744.0     NaN     NaN     NaN     NaN     NaN   
2021-02-01     1317.0      NaN     NaN     NaN     NaN     NaN     NaN   

cohort_index      8       9       10      11      12      13     14  
cohort_month                                                         
2020-01-01    4278.0  4110.0  4240.0  3831.0  3791.0  3632.0  761.0  
2020-02-01    2876.0  2633.0  2172.0  1930.0  1595.0   309.0    NaN  
2020-03-01    5009.0  4595.0  4604.0  4414.0   899.0     NaN    NaN  
2020-04-01    3356.0  3358.0  3158.0   605.0     NaN     NaN    NaN  
2020-05-01    5811.0  5598.0  1137.0     NaN     NaN     NaN    NaN  
2020-06-01    4241.0   808.0     NaN     NaN     NaN     NaN    NaN  
2020-07-01    1224.0     NaN     NaN     NaN     NaN     NaN    NaN  
2020-08-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN  
2020-09-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN  
2020-10-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN  
2020-11-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN  
2020-12-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN  
2021-01-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN  
2021-02-01       NaN     NaN     NaN     NaN     NaN     NaN    NaN

cohort_index         1          2         3         4         5         6   \
cohort_month                                                                 
2020-01-01     662180.0   642990.0  608410.0  565650.0  547080.0  483840.0   
2020-02-01     536310.0   594210.0  445590.0  450500.0  409640.0  366150.0   
2020-03-01     873000.0   797300.0  680220.0  666890.0  646610.0  546210.0   
2020-04-01     648770.0   667480.0  526090.0  503450.0  482350.0  378620.0   
2020-05-01     992030.0   870090.0  737430.0  740850.0  699940.0  633220.0   
2020-06-01     718980.0   759810.0  563430.0  559150.0  545300.0  472750.0   
2020-07-01    1106840.0   913300.0  739440.0  735260.0  661290.0  634590.0   
2020-08-01     844520.0   800350.0  607090.0  586730.0  603010.0  517030.0   
2020-09-01    1164460.0   979070.0  784700.0  778600.0  710410.0  140880.0   
2020-10-01     945800.0   852960.0  693990.0  645470.0  129240.0       NaN   
2020-11-01    1215110.0  1012760.0  782690.0  158940.0       NaN       NaN   
2020-12-01    1038060.0   947310.0  141740.0       NaN       NaN       NaN   
2021-01-01    1319750.0   268700.0       NaN       NaN       NaN       NaN   
2021-02-01      76840.0        NaN       NaN       NaN       NaN       NaN   

cohort_index        7         8         9         10        11        12  \
cohort_month                                                               
2020-01-01    455490.0  412040.0  407790.0  428040.0  382000.0  371740.0   
2020-02-01    300350.0  278920.0  251840.0  212570.0  194800.0  156240.0   
2020-03-01    542030.0  486430.0  461720.0  465630.0  447040.0   90390.0   
2020-04-01    354250.0  332740.0  329710.0  314460.0   61750.0       NaN   
2020-05-01    584100.0  572410.0  568870.0  115380.0       NaN       NaN   
2020-06-01    453890.0  411240.0   78310.0       NaN       NaN       NaN   
2020-07-01    604380.0  124890.0       NaN       NaN       NaN       NaN   
2020-08-01     94600.0       NaN       NaN       NaN       NaN       NaN   
2020-09-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-10-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-11-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-12-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-01-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-02-01         NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        13       14  
cohort_month                     
2020-01-01    360570.0  79310.0  
2020-02-01     31080.0      NaN  
2020-03-01         NaN      NaN  
2020-04-01         NaN      NaN  
2020-05-01         NaN      NaN  
2020-06-01         NaN      NaN  
2020-07-01         NaN      NaN  
2020-08-01         NaN      NaN  
2020-09-01         NaN      NaN  
2020-10-01         NaN      NaN  
2020-11-01         NaN      NaN  
2020-12-01         NaN      NaN  
2021-01-01         NaN      NaN  
2021-02-01         NaN      NaN

In [43]:
display(relation_id)
display(relation_sum)


cohort_index   1         2         3         4         5         6         7   \
cohort_month                                                                    
2020-01-01    1.0  0.817991  0.710326  0.664098  0.617470  0.513008  0.465980   
2020-02-01    1.0  0.946827  0.657568  0.660877  0.585135  0.494387  0.389106   
2020-03-01    1.0  0.740273  0.616805  0.593263  0.560939  0.454201  0.426495   
2020-04-01    1.0  0.861899  0.619579  0.594587  0.569890  0.405611  0.365406   
2020-05-01    1.0  0.721266  0.575090  0.556401  0.515819  0.441129  0.397544   
2020-06-01    1.0  0.865446  0.600938  0.584130  0.556794  0.458422  0.416667   
2020-07-01    1.0  0.671959  0.495796  0.486447  0.423590  0.392368  0.365614   
2020-08-01    1.0  0.781543  0.538044  0.525338  0.515307  0.427478  0.077129   
2020-09-01    1.0  0.683426  0.497497  0.486812  0.440301  0.083966       NaN   
2020-10-01    1.0  0.740807  0.529091  0.493517  0.102866       NaN       NaN   
2020-11-01    1.0  0.667655  0.466141  0.096574       NaN       NaN       NaN   
2020-12-01    1.0  0.751417  0.098782       NaN       NaN       NaN       NaN   
2021-01-01    1.0  0.182350       NaN       NaN       NaN       NaN       NaN   
2021-02-01    1.0       NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        8         9         10        11        12        13  \
cohort_month                                                               
2020-01-01    0.428057  0.411247  0.424255  0.383330  0.379328  0.363418   
2020-02-01    0.339832  0.311119  0.256647  0.228052  0.188467  0.036512   
2020-03-01    0.379182  0.347843  0.348524  0.334141  0.068055       NaN   
2020-04-01    0.331522  0.331720  0.311963  0.059765       NaN       NaN   
2020-05-01    0.387865  0.373648  0.075891       NaN       NaN       NaN   
2020-06-01    0.375177  0.071479       NaN       NaN       NaN       NaN   
2020-07-01    0.071970       NaN       NaN       NaN       NaN       NaN   
2020-08-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-09-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-10-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-11-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-12-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-01-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-02-01         NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        14  
cohort_month            
2020-01-01    0.076146  
2020-02-01         NaN  
2020-03-01         NaN  
2020-04-01         NaN  
2020-05-01         NaN  
2020-06-01         NaN  
2020-07-01         NaN  
2020-08-01         NaN  
2020-09-01         NaN  
2020-10-01         NaN  
2020-11-01         NaN  
2020-12-01         NaN  
2021-01-01         NaN  
2021-02-01         NaN

cohort_index   1         2         3         4         5         6         7   \
cohort_month                                                                    
2020-01-01    1.0  0.971020  0.918799  0.854224  0.826180  0.730677  0.687864   
2020-02-01    1.0  1.107960  0.830844  0.839999  0.763812  0.682721  0.560031   
2020-03-01    1.0  0.913288  0.779175  0.763906  0.740676  0.625670  0.620882   
2020-04-01    1.0  1.028839  0.810904  0.776007  0.743484  0.583597  0.546033   
2020-05-01    1.0  0.877080  0.743355  0.746802  0.705563  0.638307  0.588793   
2020-06-01    1.0  1.056789  0.783652  0.777699  0.758436  0.657529  0.631297   
2020-07-01    1.0  0.825142  0.668064  0.664288  0.597458  0.573335  0.546041   
2020-08-01    1.0  0.947698  0.718858  0.694750  0.714027  0.612218  0.112016   
2020-09-01    1.0  0.840793  0.673875  0.668636  0.610077  0.120983       NaN   
2020-10-01    1.0  0.901840  0.733760  0.682459  0.136646       NaN       NaN   
2020-11-01    1.0  0.833472  0.644131  0.130803       NaN       NaN       NaN   
2020-12-01    1.0  0.912577  0.136543       NaN       NaN       NaN       NaN   
2021-01-01    1.0  0.203599       NaN       NaN       NaN       NaN       NaN   
2021-02-01    1.0       NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        8         9         10        11        12        13  \
cohort_month                                                               
2020-01-01    0.622248  0.615830  0.646410  0.576882  0.561388  0.544520   
2020-02-01    0.520072  0.469579  0.396357  0.363223  0.291324  0.057952   
2020-03-01    0.557194  0.528889  0.533368  0.512073  0.103540       NaN   
2020-04-01    0.512878  0.508208  0.484702  0.095180       NaN       NaN   
2020-05-01    0.577009  0.573440  0.116307       NaN       NaN       NaN   
2020-06-01    0.571977  0.108918       NaN       NaN       NaN       NaN   
2020-07-01    0.112835       NaN       NaN       NaN       NaN       NaN   
2020-08-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-09-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-10-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-11-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-12-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-01-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-02-01         NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        14  
cohort_month            
2020-01-01    0.119771  
2020-02-01         NaN  
2020-03-01         NaN  
2020-04-01         NaN  
2020-05-01         NaN  
2020-06-01         NaN  
2020-07-01         NaN  
2020-08-01         NaN  
2020-09-01         NaN  
2020-10-01         NaN  
2020-11-01         NaN  
2020-12-01         NaN  
2021-01-01         NaN  
2021-02-01         NaN

In [45]:
display(relation_sum - relation_id)

cohort_index   1         2         3         4         5         6         7   \
cohort_month                                                                    
2020-01-01    0.0  0.153029  0.208472  0.190125  0.208710  0.217670  0.221885   
2020-02-01    0.0  0.161133  0.173276  0.179122  0.178677  0.188333  0.170925   
2020-03-01    0.0  0.173015  0.162370  0.170643  0.179737  0.171469  0.194387   
2020-04-01    0.0  0.166941  0.191325  0.181420  0.173593  0.177986  0.180628   
2020-05-01    0.0  0.155815  0.168264  0.190401  0.189744  0.197178  0.191249   
2020-06-01    0.0  0.191343  0.182714  0.193569  0.201642  0.199107  0.214630   
2020-07-01    0.0  0.153183  0.172268  0.177841  0.173867  0.180967  0.180427   
2020-08-01    0.0  0.166156  0.180814  0.169412  0.198720  0.184740  0.034887   
2020-09-01    0.0  0.157367  0.176377  0.181824  0.169775  0.037017       NaN   
2020-10-01    0.0  0.161032  0.204669  0.188942  0.033780       NaN       NaN   
2020-11-01    0.0  0.165817  0.177990  0.034229       NaN       NaN       NaN   
2020-12-01    0.0  0.161160  0.037761       NaN       NaN       NaN       NaN   
2021-01-01    0.0  0.021250       NaN       NaN       NaN       NaN       NaN   
2021-02-01    0.0       NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        8         9         10        11        12        13  \
cohort_month                                                               
2020-01-01    0.194191  0.204583  0.222156  0.193552  0.182061  0.181102   
2020-02-01    0.180240  0.158460  0.139710  0.135171  0.102857  0.021440   
2020-03-01    0.178011  0.181046  0.184844  0.177933  0.035485       NaN   
2020-04-01    0.181356  0.176488  0.172739  0.035415       NaN       NaN   
2020-05-01    0.189143  0.199792  0.040416       NaN       NaN       NaN   
2020-06-01    0.196800  0.037439       NaN       NaN       NaN       NaN   
2020-07-01    0.040864       NaN       NaN       NaN       NaN       NaN   
2020-08-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-09-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-10-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-11-01         NaN       NaN       NaN       NaN       NaN       NaN   
2020-12-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-01-01         NaN       NaN       NaN       NaN       NaN       NaN   
2021-02-01         NaN       NaN       NaN       NaN       NaN       NaN   

cohort_index        14  
cohort_month            
2020-01-01    0.043625  
2020-02-01         NaN  
2020-03-01         NaN  
2020-04-01         NaN  
2020-05-01         NaN  
2020-06-01         NaN  
2020-07-01         NaN  
2020-08-01         NaN  
2020-09-01         NaN  
2020-10-01         NaN  
2020-11-01         NaN  
2020-12-01         NaN  
2021-01-01         NaN  
2021-02-01         NaN

### Вывод.
1) Сразу заметим, что в последний месяц для каждой когорты у нас крайне низкие показатели, это связано с тем, что  последний месяц у нас не полный.
2) Можно заметить, что покупатели с каждым месяцем все меньше и меньше делают покупок. Однако не смотря на это уменьшение, итоговая сумма операций падает не так сильно. Например, возьмем первую строку и второй столбец. Видна сильная просадка в количестве - порядка 18%, однако в сумме операций просадка всего 3%. И так в целом можно сказать про все месяца. То есть вывод, если сильно проседает количество, то это не значит, что сумма покупок также сильно уменьшится, она естественно просядет, но не так сильно. Почему так происходит. Я думаю, что это особенность бизнеса и ответить на данный вопрос тяжело без знания самого бизнеса, так как если это магазин какой-нибудь техники, то это нормальная ситуация (люди не так часто покупают технику), а если продуктовый магазин, то это странно (люди часто ходят в магазины и если перестали ходить, то значит есть какие-то причины)